### PR TITLE
feat: [CDS-54459]: Adding Instance Sync V2 proto for NG

### PR DESCRIPTION
**Pull request by namantalaycha10**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger utAll`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/45463)

### CODE REVIEW

**Review by abosii-harness — CHANGES_REQUESTED** [review]
> Reviewed 4 of 11 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45463)* status: 11 file(s) to review. 1 change requests. 0 of 2 approvals obtained. Waiting on team(s): k8s-team. 

___
*[`126-instance/src/main/java/io/harness/repositories/instanceSyncPerpetualTaskMapping/InstanceSyncPerpetualTaskMappingRepositoryCustomImpl.java` line 2 at r1](https://reviewable.io/reviews/harness/harness-core/45463#-NR7jH_TAXOWYCmmTxwf:-NR7jH_UAPihU_EjPe2K:b9hoq2n) ([raw file](https://github.com/harness/harness-core/blob/664fe1466b3b05a4db2ad897daed7f45914b9f8c/126-instance/src/main/java/io/harness/repositories/instanceSyncPerpetualTaskMapping/InstanceSyncPerpetualTaskMappingRepositoryCustomImpl.java#L2)):*
> ```Java
> /*
> <<<<<<< HEAD
> ```

merge conflict
___
*[`126-instance/src/main/java/io/harness/service/instancesyncperpetualtask/instancesyncperpetualtaskhandler/k8s/K8SInstanceSyncPerpetualTaskHandler.java` line 83 at r1](https://reviewable.io/reviews/harness/harness-core/45463#-NR7jQ2g5CHF2WzEC_mN:-NR7jQ2h6L0Cs6u9nJEE:b-n65x4t) ([raw file](https://github.com/harness/harness-core/blob/664fe1466b3b05a4db2ad897daed7f45914b9f8c/126-instance/src/main/java/io/harness/service/instancesyncperpetualtask/instancesyncperpetualtaskhandler/k8s/K8SInstanceSyncPerpetualTaskHandler.java#L83)):*
> ```Java
> 
>     List<K8sDeploymentReleaseData> deploymentReleaseList =
>         populateDeploymentReleaseList(infrastructureMappingDTO, deploymentInfoDTOList, infrastructureOutcome);
> ```

We shouldn't need this it may end up in duplicate capabilities. You should create capabilities only based on the connector. Either duplicate the logic, create a helper, etc.

**Review by namantalaycha10 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45463)* status: 11 file(s) to review. 1 change requests. 2. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`126-instance/src/main/java/io/harness/service/instancesyncperpetualtask/instancesyncperpetualtaskhandler/k8s/K8SInstanceSyncPerpetualTaskHandler.java` line 83 at r1](https://reviewable.io/reviews/harness/harness-core/45463#-NR7jQ2g5CHF2WzEC_mN:-NRWsj1m2g4hjOPK7QCd:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/664fe1466b3b05a4db2ad897daed7f45914b9f8c/126-instance/src/main/java/io/harness/service/instancesyncperpetualtask/instancesyncperpetualtaskhandler/k8s/K8SInstanceSyncPerpetualTaskHandler.java#L83)):*
<details><summary><i>Previously, abosii-harness (Alexandru Bosii) wrote…</i></summary><blockquote>

We shouldn't need this it may end up in duplicate capabilities. You should create capabilities only based on the connector. Either duplicate the logic, create a helper, etc.
</blockquote></details>

Done.
___
*[`126-instance/src/main/java/io/harness/repositories/instanceSyncPerpetualTaskMapping/InstanceSyncPerpetualTaskMappingRepositoryCustomImpl.java` line 2 at r1](https://reviewable.io/reviews/harness/harness-core/45463#-NR7jH_TAXOWYCmmTxwf:-NRWskj73SV3fbmMFjYz:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/664fe1466b3b05a4db2ad897daed7f45914b9f8c/126-instance/src/main/java/io/harness/repositories/instanceSyncPerpetualTaskMapping/InstanceSyncPerpetualTaskMappingRepositoryCustomImpl.java#L2)):*
<details><summary><i>Previously, abosii-harness (Alexandru Bosii) wrote…</i></summary><blockquote>

merge conflict
</blockquote></details>

Done.

**Review by abosii-harness — CHANGES_REQUESTED** [review]
> Reviewed 2 of 11 files at r1, 5 of 7 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45463)* status: 10 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`126-instance/src/main/java/io/harness/service/instancesyncperpetualtask/instancesyncperpetualtaskhandler/k8s/K8SInstanceSyncPerpetualTaskHandler.java` line 209 at r3](https://reviewable.io/reviews/harness/harness-core/45463#-NRXNOd52fcuVEGHDBsI:-NRXNOd52fcuVEGHDBsJ:b-g664ff) ([raw file](https://github.com/harness/harness-core/blob/77cac694d9616885e7ea0af892ab4b5f570387c9/126-instance/src/main/java/io/harness/service/instancesyncperpetualtask/instancesyncperpetualtaskhandler/k8s/K8SInstanceSyncPerpetualTaskHandler.java#L209)):*
> ```Java
>     }
> 
>     return K8sTaskCapabilityHelper.fetchRequiredExecutionCapabilities(connectorInfoDTO.getConnectorConfig(), null,
> ```

What about GKE, AKS, etc? We should have a Utility class that based on the type of connector get capabilities

File: 126-instance/src/main/java/io/harness/service/instancesyncperpetualtask/instancesyncperpetualtaskhandler/k8s/K8SInstanceSyncPerpetualTaskHandler.java [hunk 1]
```diff
@@ -51,6 +56,7 @@
 @OwnedBy(HarnessTeam.CDP)
 public class K8SInstanceSyncPerpetualTaskHandler extends InstanceSyncPerpetualTaskHandler {
   @Inject private K8sEntityHelper k8sEntityHelper;
+  @Inject private CDFeatureFlagHelper cdFeatureFlagHelper;
```
Comment (abosii-harness) [inline 126-instance/src/main/java/io/harness/service/instancesyncperpetualtask/instancesyncperpetualtaskhandler/k8s/K8SInstanceSyncPerpetualTaskHandler.java:59]: Should use accountClient, this cdFeatureFlagHelper is not working properly in NG

**Review by abosii-harness — CHANGES_REQUESTED** [review]

**Review by namantalaycha10 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45463)* status: 12 file(s) to review. 1 change requests. 2. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`126-instance/src/main/java/io/harness/service/instancesyncperpetualtask/instancesyncperpetualtaskhandler/k8s/K8SInstanceSyncPerpetualTaskHandler.java` line 209 at r3](https://reviewable.io/reviews/harness/harness-core/45463#-NRXNOd52fcuVEGHDBsI:-NRatgUU13c69L9LKiIt:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/77cac694d9616885e7ea0af892ab4b5f570387c9/126-instance/src/main/java/io/harness/service/instancesyncperpetualtask/instancesyncperpetualtaskhandler/k8s/K8SInstanceSyncPerpetualTaskHandler.java#L209)):*
<details><summary><i>Previously, abosii-harness (Alexandru Bosii) wrote…</i></summary><blockquote>

What about GKE, AKS, etc? We should have a Utility class that based on the type of connector get capabilities
</blockquote></details>

Done.

File: 126-instance/src/main/java/io/harness/service/instancesync/InstanceSyncServiceImpl.java [hunk 2]
```diff
@@ -210,32 +179,100 @@ public void processInstanceSyncForNewDeployment(DeploymentEvent deploymentEvent)
...
+                instanceSyncPerpetualTaskInfoDTO.getDeploymentInfoDetailsDTOList())
+            || instanceSyncPerpetualTaskInfoDTO.getConnectorIdentifier() == null) {
+          addNewDeploymentInfoToInstanceSyncPerpetualTaskInfoRecord(
+              instanceSyncPerpetualTaskInfoDTO, deploymentSummaryDTO);
+          instanceSyncPerpetualTaskInfoDTO =
+              instanceSyncPerpetualTaskInfoService.updateDeploymentInfoListAndConnectorId(
+                  instanceSyncPerpetualTaskInfoDTO, infrastructureMappingDTO.getConnectorRef());
+
+          // Reset perpetual task to update the execution bundle with the latest information
+          instanceSyncPerpetualTaskService.resetPerpetualTask(infrastructureMappingDTO.getAccountIdentifier(),
```
Comment (abosii-harness) [inline 126-instance/src/main/java/io/harness/service/instancesync/InstanceSyncServiceImpl.java:265]: We will need to handle resetPerpetualTask with a different method

File: 126-instance/src/main/java/io/harness/service/instancesync/InstanceSyncServiceImpl.java [hunk 3]
```diff
@@ -210,32 +179,100 @@ public void processInstanceSyncForNewDeployment(DeploymentEvent deploymentEvent)
...
+    InstanceSyncPerpetualTaskInfoDTO instanceSyncPerpetualTaskInfoDTO;
+    if (instanceSyncPerpetualTaskInfoDTOOptional.isEmpty()) {
+      // no existing perpetual task info record found for given infrastructure mapping id
+      // so create a new perpetual task and instance sync perpetual task info record
+      String perpetualTaskId = instanceSyncPerpetualTaskService.createPerpetualTask(infrastructureMappingDTO,
+          abstractInstanceSyncHandler, Collections.singletonList(deploymentSummaryDTO.getDeploymentInfoDTO()),
+          deploymentEvent.getInfrastructureOutcome());
+      instanceSyncPerpetualTaskInfoDTO = instanceSyncPerpetualTaskInfoService.save(
+          prepareInstanceSyncPerpetualTaskInfoDTO(deploymentSummaryDTO, perpetualTaskId,
+              instanceSyncPerpetualTaskService.isInstanceSyncV2Enabled() ? infrastructureMappingDTO.getConnectorRef()
```
Comment (abosii-harness) [inline 126-instance/src/main/java/io/harness/service/instancesync/InstanceSyncServiceImpl.java:196]: either always set as connector ref either null

**Review by abosii-harness — CHANGES_REQUESTED** [review]

**Review by achyuth-harness — APPROVED** [review]
> Reviewed 3 of 11 files at r1, 1 of 3 files at r4, 1 of 4 files at r5, 4 of 4 files at r6, 3 of 3 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45463)* status: :shipit: complete! ✓

**Pull request merged**